lib/Frontend/ParseableInterfaceModuleLoader.cpp

```diff
@@ -0,0 +1,1022 @@
+//===-- ParseableInterfaceModuleLoader.cpp - Loads .swiftinterface files --===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2018 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#define DEBUG_TYPE "textual-module-interface"
+#include "swift/Frontend/ParseableInterfaceModuleLoader.h"
+#include "swift/AST/ASTContext.h"
+#include "swift/AST/DiagnosticsFrontend.h"
+#include "swift/AST/DiagnosticsSema.h"
+#include "swift/AST/FileSystem.h"
+#include "swift/AST/Module.h"
+#include "swift/AST/ProtocolConformance.h"
+#include "swift/Basic/STLExtras.h"
+#include "swift/Frontend/Frontend.h"
+#include "swift/Frontend/ParseableInterfaceSupport.h"
+#include "swift/Frontend/PrintingDiagnosticConsumer.h"
+#include "swift/SILOptimizer/PassManager/Passes.h"
+#include "swift/Serialization/ModuleFormat.h"
+#include "swift/Serialization/SerializationOptions.h"
+#include "swift/Serialization/Validation.h"
+#include "clang/Basic/Module.h"
+#include "clang/Lex/Preprocessor.h"
+#include "clang/Lex/PreprocessorOptions.h"
+#include "clang/Lex/HeaderSearch.h"
+#include "clang/Frontend/CompilerInstance.h"
+#include "llvm/ADT/Hashing.h"
+#include "llvm/ADT/StringSet.h"
+#include "llvm/Support/xxhash.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/CrashRecoveryContext.h"
+#include "llvm/Support/Path.h"
+#include "llvm/Support/Errc.h"
+#include "llvm/Support/Regex.h"
+#include "llvm/Support/StringSaver.h"
+#include "llvm/Support/YAMLTraits.h"
+
+using namespace swift;
+using FileDependency = SerializationOptions::FileDependency;
+
+/// Extract the specified-or-defaulted -module-cache-path that winds up in
+/// the clang importer, for reuse as the .swiftmodule cache path when
+/// building a ParseableInterfaceModuleLoader.
+std::string
+swift::getModuleCachePathFromClang(const clang::CompilerInstance &Clang) {
+  if (!Clang.hasPreprocessor())
+    return "";
+  std::string SpecificModuleCachePath = Clang.getPreprocessor()
+    .getHeaderSearchInfo()
+    .getModuleCachePath();
+
+  // The returned-from-clang module cache path includes a suffix directory
```

Comment:
If `-fdisable-module-hash` is set, then you would **not** want the parent path.  It's a `-cc1` option though, so maybe we don't care.